subsys/net/lib/tls_credentials/tls_credentials_shell.c

```diff
@@ -0,0 +1,813 @@
+	}
+
+	shell_fprintf(
+		sh, SHELL_NORMAL,
+		"Stored %d bytes.\n", res
+	);
+
+	return 0;
+}
+
+/* Adds a credential to the credential store */
+static int tls_cred_cmd_add(const struct shell *sh, size_t argc, char *argv[])
+{
+	int err = 0;
+	sec_tag_t sectag;
+	enum cred_storage_fmt format;
+	bool terminated;
+	enum tls_credential_type type;
+	void *cred_copy = NULL;
+	void *cred_chosen = NULL;
+	bool keep_copy = false;
+
+	/* Lock credentials so that we can interact with them directly.
+	 * Mainly this is required by credential_get.
+	 */
+
+	credentials_lock();
+
+	err = shell_parse_cred_sectag(sh, argv[1], &sectag, false);
+	if (err) {
+		goto cleanup;
+	}
+
+	err = shell_parse_cred_type(sh, argv[2], &type, false);
+	if (err) {
+		goto cleanup;
+	}
+
+	err = shell_parse_cred_backend(sh, argv[3]);
+	if (err) {
+		goto cleanup;
+	}
+
+	err = shell_parse_cred_storage_format(sh, argv[4], &format, &terminated);
+	if (err) {
+		goto cleanup;
+	}
+
+	if (argc == 6) {
+		/* Credential was passed, clear credential buffer and then use the passed-in
+		 * credential.
+		 */
+		shell_clear_cred_buf(sh);
+		err = shell_write_cred_buf(sh, argv[5]);
+
+		if (err) {
+			goto cleanup;
+		}
+	}
+
+	/* Make sure the credential buffer isn't empty. */
+	if (cred_written == 0) {
+		shell_fprintf(
+			sh, SHELL_ERROR,
+			"Please provide a credential to add.\n"
+		);
+		err = -ENOENT;
+		goto cleanup;
+	}
+
+	/* Check whether a credential of this type and sectag already exists. */
+	if (credential_get(sectag, type)) {
+		shell_fprintf(
+			sh, SHELL_ERROR,
+			"TLS credential with sectag %d and type %s already exists.\n", sectag,
+			cred_type_name(type)
+		);
+		err = -EEXIST;
+		goto cleanup;
+	}
+
+	/* If binary format was specified, decode from base64. */
+	if (format == CRED_STORAGE_FMT_BINARY) {
+		/* base64_decode can handle in-place operation.
+		 * Pass &cred_written as olen so that it is updated to match the size of the base64
+		 * encoding.
+		 *
+		 * We use sizeof(cred_buf)-1 since we want to keep room for a NULL terminator.
+		 * Though, technically, this is not actually needed since the output of
+		 * base64_decode is always shorter than its input.
+		 */
+		err = base64_decode(cred_buf, sizeof(cred_buf)-1, &cred_written,
+				    cred_buf, cred_written);
+
+		if (err) {
+			shell_fprintf(
+				sh, SHELL_ERROR,
+				"Could not decode input from base64, error: %d\n",
+				err
+			);
+			err = -EINVAL;
+			goto cleanup;
+		}
+	}
+
+	/* If NULL termination was requested, append one.
+	 * We are always guaranteed to have room in the buffer for this.
+	 */
+	if (terminated) {
+		cred_buf[cred_written] = 0;
+		cred_written += 1;
+	}
+
+	/* Default to using cred_buf directly. */
+	cred_chosen = cred_buf;
+
+	/* If the currently active TLS Credentials backend stores credentials by reference,
+	 * copy the incoming credentials off to the heap, and then use this copy instead.
+	 */
+	if (COPY_CREDENTIALS_TO_HEAP) {
+		cred_copy = k_malloc(cred_written);
```

Comment:
For the volatile storage, we allocate the memory here, but do not free it on `delete` operation. Maybe the shell module should keep a list of own allocations (just a memory pointer) so that it could be verified whether to release the allocated memory or not when deleting credentials. Currently, we have a memory leak, maybe not criticial, but still.

Reply:
Maybe you're right. I noticed this leak as well responding to Jukka's comments. My initial instinct was to keep the leak, but warn the user, just for simplicity's sake, since this is really only a concern for the volatile back-end. But I might as well go the full mile and keep an allocation record, that way there are fewer things for the user to worry about. I'll do that.

Reply:
This still technically leaves a leak if the application itself directly calls `tls_credential_delete` or if the user deletes any credentials that happen to be heap-allocated and were not created by the credentials shell, but at that point I think the cost of preventing that kind of leak far outweighs the benefits